arangod/Aql/Functions.cpp

```diff
@@ -1604,6 +1604,74 @@ AqlValue Functions::Sleep(arangodb::aql::Query* query,
   return AqlValue(AqlValueHintNull());
 }
 
+#include "GeneralServer/AuthenticationFeature.h"
+#include "RestServer/FeatureCacheFeature.h"
+#include "V8Server/v8-collection.h"
+
+/// @brief function COLLECTIONS
+AqlValue Functions::Collections(arangodb::aql::Query* query,
+                          transaction::Methods* trx,
+                          VPackFunctionParameters const& parameters) {
+
+  transaction::BuilderLeaser builder(trx);
+  builder->openArray();
+
+  TRI_vocbase_t* vocbase = query->vocbase();
+
+  std::vector<LogicalCollection*> colls;
+
+  // clean memory
+  std::function<void()> cleanup;
+
+  // if we are a coordinator, we need to fetch the collection info from the
+  // agency
+  if (ServerState::instance()->isCoordinator()) {
+    cleanup = [&colls]() {
+      for (auto& it : colls) {
+        if (it != nullptr) {
+          delete it;
+        }
+      }
+    };
+    colls = GetCollectionsCluster(vocbase);
+  } else {
+    colls = vocbase->collections(false);
+  }
+
+  // make sure memory is cleaned up
+  TRI_DEFER(cleanup());
+
+  std::sort(colls.begin(), colls.end(), [](LogicalCollection* lhs, LogicalCollection* rhs) -> bool {
+    return basics::StringUtils::tolower(lhs->name()) < basics::StringUtils::tolower(rhs->name());
+  });
+
+  AuthenticationFeature* auth = FeatureCacheFeature::instance()->authenticationFeature();
+
+  size_t const n = colls.size();
+  for (size_t i = 0; i < n; ++i) {
+    auto& collection = colls[i];
+
+    if (auth->isActive() && ExecContext::CURRENT != nullptr) {
+      AuthLevel level = auth->canUseCollection(ExecContext::CURRENT->user(),
+                             vocbase->name(), collection->name());
+      if (level == AuthLevel::NONE) {
+        continue;
+      }
+    }
+
+    builder->openObject();
+    builder->add("_id", VPackValue(collection->cid_as_string()));
+    builder->add("name", VPackValue(collection->name()));
+    builder->close();
+
+    collection = nullptr;
```

Comment:
this will produce a memleak. please delete this line.

Reply:
done